core/src/main/java/org/apache/iceberg/ManifestMergeManager.java

```diff
@@ -0,0 +1,190 @@
+      Iterables.addAll(merged, mergeGroup(first, specId, groups.get(specId)));
+    }
+
+    return merged;
+  }
+
+  void cleanUncommitted(Set<ManifestFile> committed) {
+    // iterate over a copy of entries to avoid concurrent modification
+    List<Map.Entry<List<ManifestFile>, ManifestFile>> entries =
+        Lists.newArrayList(mergedManifests.entrySet());
+
+    for (Map.Entry<List<ManifestFile>, ManifestFile> entry : entries) {
+      // delete any new merged manifests that aren't in the committed list
+      ManifestFile merged = entry.getValue();
+      if (!committed.contains(merged)) {
+        deleteFile(merged.path());
+        // remove the deleted file from the cache
+        mergedManifests.remove(entry.getKey());
+      }
+    }
+  }
+
+  private ListMultimap<Integer, ManifestFile> groupBySpec(ManifestFile first, Iterator<ManifestFile> remaining) {
+    ListMultimap<Integer, ManifestFile> groups = Multimaps.newListMultimap(
+        Maps.newTreeMap(Comparator.<Integer>reverseOrder()),
+        Lists::newArrayList);
+    groups.put(first.partitionSpecId(), first);
+    remaining.forEachRemaining(manifest -> groups.put(manifest.partitionSpecId(), manifest));
+    return groups;
+  }
+
+  @SuppressWarnings("unchecked")
+  private Iterable<ManifestFile> mergeGroup(ManifestFile first, int specId, List<ManifestFile> group) {
+    // use a lookback of 1 to avoid reordering the manifests. using 1 also means this should pack
+    // from the end so that the manifest that gets under-filled is the first one, which will be
+    // merged the next time.
+    ListPacker<ManifestFile> packer = new ListPacker<>(targetSizeBytes, 1, false);
+    List<List<ManifestFile>> bins = packer.packEnd(group, ManifestFile::length);
+
+    // process bins in parallel, but put results in the order of the bins into an array to preserve
+    // the order of manifests and contents. preserving the order helps avoid random deletes when
+    // data files are eventually aged off.
+    List<ManifestFile>[] binResults = (List<ManifestFile>[])
+        Array.newInstance(List.class, bins.size());
+
+    Tasks.range(bins.size())
+        .stopOnFailure().throwFailureWhenFinished()
+        .executeWith(ThreadPools.getWorkerPool())
+        .run(index -> {
+          List<ManifestFile> bin = bins.get(index);
+          List<ManifestFile> outputManifests = Lists.newArrayList();
+          binResults[index] = outputManifests;
+
+          if (bin.size() == 1) {
+            // no need to rewrite
+            outputManifests.add(bin.get(0));
+            return;
+          }
+
+          // if the bin has the first manifest (the new data files or an appended manifest file) then only merge it
+          // if the number of manifests is above the minimum count. this is applied only to bins with an in-memory
+          // manifest so that large manifests don't prevent merging older groups.
+          if (bin.contains(first) && bin.size() < minCountToMerge) {
```

Comment:
This introduces a slight behavior change. Before, the first new manifest was used to identify the starting bin to apply the minimum merge count. As a result, the min count had no effect if there were no added files or appended manifests.

Now, this uses the first manifest in the manifest list passed to merge to identify the initial bin. This will result in some operations that did not add files or manifests not being merged, like the delete operations. This is what required the change to the v2 sequence number transaction test.

Reply:
Could you please help me to understand the following case?

If we append 5 manifests on a table that has 1 existing manifests, and they are packed into 2 bins? Assume we set minCountToMerge to 4.

[(m1, m2, m3, m4), (m5, m6)], m1-m5 are appended manifests, m6 is an existing table manifest.

The `first` is m1, right?  then both bins are getting merged?

Reply:
`m1` is used to detect the first bin, so it would not be merged because it only has 2 (< `minCount`) manifests. The other bin would be merged because it isn't the first.

Overall, the behavior is the same as before. The difference is that now we can detect the first bin even if there are no appended manifests. In your example, that commit would end with `[m1, m2, m7]`, where `m7` is the merged bin. A delete commit after that would produce bins `[(m1, m2), (m7)]` but this time the first bin, `(m1, m2)` would get merged because there was no appended manifest used to detect that it was the first bin. Now, we use the first manifest, `m1`, to identify the first bin and it would not get merged.

Reply:
Seems reasonable!

Reply:
Thanks for your elaboration!